File: app/models/speed_grader/submission.rb
#
# Copyright (C) 2018 - present Instructure, Inc.
#
# This file is part of Canvas.
#
# Canvas is free software: you can redistribute it and/or modify it under
# the terms of the GNU Affero General Public License as published by the Free
# Software Foundation, version 3 of the License.
#
# Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
# details.
#
# You should have received a copy of the GNU Affero General Public License along
# with this program. If not, see <http://www.gnu.org/licenses/>.
#

class SpeedGrader::Submission
  def initialize(submission:, current_user:, provisional_grade:)
    @submission = submission
    @current_user = current_user
    @provisional_grade = provisional_grade
  end

  def comments
    comments = if submission.assignment.grades_published?
      submission.submission_comments
    elsif grader_comments_hidden?
      (provisional_grade || submission).submission_comments
    else
      submission.all_submission_comments
    end
    comments = comments.for_groups if submission.assignment.grade_as_group?
    comments
  end

  private

  def grader_comments_hidden?
    !submission.assignment.can_view_other_grader_comments?(current_user)
  end

  attr_reader :submission, :current_user, :provisional_grade
end

File: app/coffeescripts/models/Submission.js
//
// Copyright (C) 2011 - present Instructure, Inc.
//
// This file is part of Canvas.
//
// Canvas is free software: you can redistribute it and/or modify it under
// the terms of the GNU Affero General Public License as published by the Free
// Software Foundation, version 3 of the License.
//
// Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
// WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
// A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
// details.
//
// You should have received a copy of the GNU Affero General Public License along
// with this program. If not, see <http://www.gnu.org/licenses/>.
//

import Backbone from 'Backbone'

export default class Submission extends Backbone.Model {
  constructor(...args) {
    super(...args)
    this.isGraded = this.isGraded.bind(this)
    this.hasSubmission = this.hasSubmission.bind(this)
    this.withoutGradedSubmission = this.withoutGradedSubmission.bind(this)
    this.present = this.present.bind(this)
  }

  isGraded() {
    return this.get('grade') != null
  }

  hasSubmission() {
    return !!this.get('submission_type')
  }

  withoutGradedSubmission() {
    return !this.hasSubmission() && !this.isGraded()
  }

  present() {
    const json = this.toJSON()
    json.submitted_or_graded = !this.withoutGradedSubmission()
    return json
  }
}
